### PR TITLE
[CLOUDTRUST-3768] Denied terms of use management

### PR DESCRIPTION
**Pull request by fperot74**

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -31,7 +31,8 @@
 
 [[constraint]]
   name = "github.com/cloudtrust/keycloak-client"
-  version = "2.4.3"
+  branch = "ct-3768"
+#version = "2.4.3"
```
Comment (fperot74) [inline Gopkg.toml:35]: To be updated before merging

File: api/tasks/swagger-api_tasks.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+- url: http://localhost:8888
+tags:
+- name: Tasks
+  description: Tasks management
+paths:
+  /tasks/denied-terms-of-use-users:
+    delete:
+      tags:
+      - Tpackage pkgtasks
+asks
```
Comment (basbeu) [inline api/tasks/swagger-api_tasks.yaml:16]: I think this word should not be here ! It brokes the swagger syntax

**Review by basbeu — APPROVED** [review]

File: api/tasks/swagger-api_tasks.yaml (hunk 2 above)
Comment (fperot74) [reply api/tasks/swagger-api_tasks.yaml:16]: Good catch ! :-)
That's what happens when you have multiple displays, you type CTRL+C but you are not watching the correct display...

File: pkg/management/component.go [hunk 3]
```diff
@@ -316,13 +316,22 @@ func (c *component) GetRequiredActions(ctx context.Context, realmName string) ([
...
+func (c *component) CreateUser(ctx context.Context, realmName string, user api.UserRepresentation, generateUsername bool, generateNameID bool, termsOfUse bool) (string, error) {
 	var accessToken = ctx.Value(cs.CtContextAccessToken).(string)
 	var ctxRealm = ctx.Value(cs.CtContextRealm).(string)
 
 	var userRep = api.ConvertToKCUser(user)
 	userRep.SetAttributeString(constants.AttrbSource, "api")
 
+	if termsOfUse {
+		var reqActions []string
+		if userRep.RequiredActions != nil {
```
Comment (harture) [inline pkg/management/component.go:328]: I don't see how it could happen

File: pkg/tasks/component.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/cloudtrust/common-service/log"
+	api "github.com/cloudtrust/keycloak-bridge/api/tasks"
+	"github.com/cloudtrust/keycloak-bridge/internal/keycloakb"
+	"github.com/cloudtrust/keycloak-client"
+)
+
+// Component interface exposes methods used by the bridge API
+type Component interface {
+	GetActions(ctx context.Context) ([]api.ActionRepresentation, error)
+	DeleteDeniedTermsOfUseUsers(ctx context.Context) error
```
Comment (harture) [inline pkg/tasks/component.go:18]: It is still the old naming

File: pkg/tasks/component.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+func (c *component) reportEvent(ctx context.Context, apiCall string, values ...string) {
+	errEvent := c.eventDBModule.ReportEvent(ctx, apiCall, "scheduled-tasks", values...)
+	if errEvent != nil {
+		//store in the logs also the event that failed to be stored in the DB
+		keycloakb.LogUnrecordedEvent(ctx, c.logger, apiCall, errEvent.Error(), values...)
+	}
+}
+
+func (c *component) GetActions(ctx context.Context) ([]api.ActionRepresentation, error) {
```
Comment (harture) [inline pkg/tasks/component.go:59]: Do we have sync with Fratt so we can really set those rights ?

File: pkg/tasks/component.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		return err
+	}
+
+	// Best effort: in case of error go on trying to remove other users
+	var finalError error
+
+	for _, user := range users {
+		err = c.usersDBModule.DeleteUserDetails(ctx, user.RealmName, user.UserID)
+		if err != nil && err != sql.ErrNoRows {
+			c.logger.Warn(ctx, "msg", "Could not delete database user", "realm", user.RealmName, "usr", user.UserID)
```
Comment (harture) [inline pkg/tasks/component.go:90]: We need to log the error else we won't be able to debug and fix the issue

File: pkg/tasks/component.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+			finalError = err
+		} else {
+			err = c.keycloakClient.DeleteUser(accessToken, user.RealmName, user.UserID)
+			if err != nil {
+				c.logger.Warn(ctx, "msg", "Could not delete keycloak user", "realm", user.RealmName, "usr", user.UserID)
```
Comment (harture) [inline pkg/tasks/component.go:95]: Would be good to log the error here too

File: pkg/tasks/component.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+				finalError = err
+			} else {
+				c.logger.Info(ctx, "msg", "Removed user without terms and conditions acceptance", "realm", user.RealmName, "usr", user.UserID)
+				//store the API call into the DB
+				c.reportEvent(ctx, "API_ACCOUNT_DELETION", database.CtEventRealmName, user.RealmName, database.CtEventUserID, user.UserID)
```
Comment (harture) [inline pkg/tasks/component.go:100]: Is it a newkind of event ? If yes, I have to add it to the list in EPR documentation

File: pkg/tasks/authorization.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	a := security.Action{
+		Name:  as,
+		Scope: scope,
+	}
+
+	actions = append(actions, a)
+	return a
+}
+
+// Actions used for authorization module
```
Comment (harture) [inline pkg/tasks/authorization.go:24]: Do we need to put authorization layer with rights for this API ?

File: cmd/keycloakb/keycloak_bridge.go [hunk 10]
```diff
@@ -979,6 +1005,13 @@ func main() {
 		communicationsSubroute.Path("/realms/{realm}/send-mail").Methods("POST").Handler(sendMailHandler)
```
Comment (harture) [inline cmd/keycloakb/keycloak_bridge.go:1005]: it seems we have an indentation issue

**Review by harture — APPROVED** [review]

File: pkg/tasks/component.go (hunk 8 above)
Comment (fperot74) [reply pkg/tasks/component.go:100]: No... it's already used in management/deleteUser

File: pkg/management/component.go (hunk 3 above)
Comment (fperot74) [reply pkg/management/component.go:328]: Currently, it's not possible... but if someone sets the RequiredActions field with one or more required actions in api.ConvertToKCUser(...), he might not see that his required actions might be ignored as in most cases, he won't set termsOfUse=true

File: pkg/tasks/component.go (hunk 4 above)
Comment (fperot74) [reply pkg/tasks/component.go:18]: We did not speak about a new name here... GetExpiredTermsOfUseAcceptance would no make sense here.
CleanUpAccordingToExpiredTermsOfUseAcceptance ?

File: pkg/tasks/component.go (hunk 5 above)
Comment (fperot74) [reply pkg/tasks/component.go:59]: I removed GetActions in both tasks and communications...
I also added the only action from tasks in the GetActions from management, as we did for actions from communications

File: cmd/keycloakb/keycloak_bridge.go (hunk 10 above)
Comment (fperot74) [reply cmd/keycloakb/keycloak_bridge.go:1005]: ???

Comment (harture) [reply cmd/keycloakb/keycloak_bridge.go:1001]: Interesting... it seems it was a bug in the diff viewer. It is now ok

File: pkg/tasks/component.go (hunk 4 above)
Comment (harture) [reply pkg/tasks/component.go]: DeleteUsersWithExpiredTermsOfUseAcceptance ?

File: pkg/management/component.go (hunk 3 above)
Comment (harture) [reply pkg/management/component.go:328]: ok

File: pkg/tasks/authorization.go (hunk 9 above)
Comment (fperot74) [reply pkg/tasks/authorization.go:23]: Discussed together by phone...

**Pull request merged**